src/include/rho/AllocationTable.hpp

```diff
@@ -0,0 +1,173 @@
+#include <limits>
+#include <map>
+
+#include "rho/AllocatorSuperblock.hpp"
+
+typedef std::uint32_t u32;
+
+using std::function;
+using std::size_t;
+
+namespace rho {
+  /*
+   * Special-purpose hashtable implementation for tracking medium and large
+   * allocations.
+   *
+   * The table uses open addressing with quadratic probing to resolve
+   * collisions. This hashtable is different from dense_hash_table by
+   * allowing asymmetric insertion and searching - allocation ranges are
+   * inserted but the lookup function searches on a pointer so the search
+   * will look for a range covering the pointer.
+   */
+  class AllocationTable {
+  public:
+    /**
+     * Hash bucket for the allocation table.
+     * The two lowest bits of the data pointer are flags indicating
+     * if the hash bucket contains a superblock (2) and whether this
+     * hash bucket is the first representing that allocation (1).
+     */
+    struct HashBucket {
+      uintptr_t data;
+      unsigned size;
+
+      /* Masks away the pointer bits from the data field. */
+      uintptr_t DataPointer() {
+        return data & ~uintptr_t{3};
+      }
+    };
+
+    /**
+     * @param num_bits the number of bits to use for hashing. The number of
+     * buckets is 2^num_bits.
+     */
+    AllocationTable(unsigned num_bits);
+
+    /**
+     * Add an allocation (large object or medium object superblock) to the
+     * allocations hashtable.
+     */
+    void insert(uintptr_t pointer, size_t size);
+
+    /*
+     * Erases all hashtable entries for a pointer. The size determines
+     * how many hash buckets are removed.
+     */
+    void erase(uintptr_t pointer, size_t size);
+
+    /*
+     * Find a bucket matching the pointer. Will find internal pointers.
+     */
+    HashBucket* search(uintptr_t candidate);
+
+    /*
+     * Remove an allocation from the hashtable.  If the allocation is in a
+     * superblock, the superblock is left as is but the superblock bitset is
+     * updated to indicate that the block is free.
+     * If the allocation is a large object (ie. not in a superblock), then
+     * the object is taken out of the hashtable and inserted into a freelist.
+     */
+    bool free_allocation(uintptr_t pointer);
+
+    /**
+     * Iterates over all live objects and calls the argument function on
+     * their pointers.
+     * It is okay to free objects during the iteration, but adding objects
+     * means they may not be found during the current iteration pass.
+     */
+    void ApplyToAllBlocks(std::function<void(void*)> fun);
+
+    /**
+     * Prints a summary of the table illustrating the bucket utilization.
+     */
+    void PrintTable();
+
+  private:
+    unsigned num_bits;
+    unsigned num_buckets;
+
+    // Mask deciding which bits of a hash key are used to index the hash table.
+    unsigned hash_mask;
+
+    HashBucket* buckets = nullptr;
+
+    // Counters logging collision statistics:
+    unsigned add_collisions = 0;
```

Comment:
These should be insert / erase / search to match the names of the functions.  Adding 'num_' would make it clearer too.
    m_num_insert_collisions